chain/epoch_manager/src/shard_assignment.rs

```diff
@@ -87,55 +87,44 @@ fn assign_with_possible_repeats<T: HasStake + Eq, I: Iterator<Item = T>>(
         let cp = cp_iter
             .next()
             .expect("cp_iter should contain enough elements to minimally fill each shard");
-        let (least_validator_count, shard_stake, shard_id) =
-            shard_index.pop().expect("shard_index should never be empty");
-
-        if result[usize::try_from(shard_id).unwrap()].contains(&cp) {
-            // `cp` is already assigned to this shard, need to assign elsewhere
-
-            // `buffer` tracks shards `cp` is already in, these will need to be pushed back into
-            // shard_index when we eventually assign `cp`.
-            buffer.push((least_validator_count, shard_stake, shard_id));
-            loop {
-                // We can still expect that there exists a shard cp has not been assigned to
-                // because we check above that there are at least `min_validators_per_shard` distinct
-                // chunk producers. This means the worst case scenario is in the end every chunk
-                // producer is assigned to every shard, in which case we would not be trying to
-                // assign a chunk producer right now.
-                let (least_validator_count, shard_stake, shard_id) =
-                    shard_index.pop().expect("shard_index should never be empty");
-                if least_validator_count >= min_validators_per_shard {
+        // Decide which shard to assign this chunk producer to.  We mustn’t
+        // assign producers to a single shard multiple times.
+        loop {
+            match shard_index.peek_mut() {
+                None => {
+                    // No shards left which don’t already contain this chunk
+                    // producer.  Skip it and move to another validator.
+                    break;
+                }
+                Some(top) if top.0 >= min_validators_per_shard => {
                     // All remaining shards have min_validators_per_shard chunk
-                    // producers assigned to them.  Don’t assign current cp to
-                    // any shard and move to next cp.
-                    buffer.push((least_validator_count, shard_stake, shard_id));
+                    // producers assigned to them.  Don’t assign current
+                    // validator to any shard and move to next cp.
```

Comment:
actually in this case - we'll probably also fail the line 86 condition - and completely exit.

Reply:
No. We may have a shard which still needs another chunk producer assigned to it but the current producer is already assigned to that shard. In this case condition on line 86 will be true, we then get to the case on line 105 and skip the shard for now and get back to looking at next shard which might turn out to have all the validators we need.